scripts/upgrade-facets.js

```diff
@@ -0,0 +1,244 @@
+const packageFile = require("../package.json");
+const readline = require("readline");
+const rl = readline.createInterface({
+  input: process.stdin,
+  output: process.stdout,
+});
+
+/**
+ * Upgrades or reamoves existing facets, or add new facets.
+ *
+ * Process:
+ *  1.  Edit scripts/config/facet-upgrade.js.
+ *  1a. Provide a list of facets that needs to be upgraded (field "addOrUpgrade") or removed completely (field "remove")
+ *  1b. Optionally you can specify which selectors should be ignored (filed "skip"). You don't have to specify "initialize()" since it's ignored by default
+ *  2. Update protocol version in package.json. If not, script will prompt you to confirm that version remains unchanged.
+ *  2. Run the appropriate npm script in package.json to upgrde facets for a given network
+ *  3. Save changes to the repo as a record of what was upgraded
+ */
+async function main() {
+  // Bail now if hardhat network
+  if (network === "hardhat") process.exit();
+
+  const chainId = (await hre.ethers.provider.getNetwork()).chainId;
+  const contractsFile = readContracts(chainId, network);
+  let contracts = contractsFile.contracts;
+
+  const divider = "-".repeat(80);
+  console.log(`${divider}\nBoson Protocol Contract Suite Upgrader\n${divider}`);
+  console.log(`⛓  Network: ${hre.network.name}\n📅 ${new Date()}`);
+
+  // Check that package.json version was updated
+  if (packageFile.version == contractsFile.protocolVersion) {
+    const answer = await new Promise((resolve) => {
+      rl.question("Protocol version has not been updated. Proceed anyway? (y/n) ", resolve);
+    });
+    switch (answer.toLowerCase()) {
+      case "y":
+      case "yes":
+        break;
+      case "n":
+      case "no":
+        process.exit(1);
+        break;
+      default:
+        break;
+    }
+  }
+
+  // Get the accounts
+  const accounts = await ethers.provider.listAccounts();
+  const admin = accounts[0];
+  console.log("🔱 Admin account: ", admin ? admin : "not found" && process.exit());
+  console.log(divider);
+
+  // Get addresses of currently deployed contracts
+  const protocolAddress = contracts.find((c) => c.name === "ProtocolDiamond").address;
+  const accessControllerAddress = contracts.find((c) => c.name === "AccessController").address;
+
+  if (!protocolAddress) {
+    return addressNotFound("ProtocolDiamond");
+  }
+
+  if (!accessControllerAddress) {
+    return addressNotFound("AccessController");
+  }
+
+  // Get AccessController abstraction
+  const accessController = await ethers.getContractAt("AccessController", accessControllerAddress);
+
+  // Check that caller has upgrader role.
+  const hasRole = await accessController.hasRole(Role.UPGRADER, admin);
+  if (!hasRole) {
+    console.log("Admin address does not have UPGRADER role");
+    process.exit(1);
+  }
+
+  // Deploy new facets
+  const deployedFacets = await deployProtocolHandlerFacets(
+    protocolAddress,
+    Facets.addOrUpgrade,
```

Comment:
In both cases we need to deploy whole new facets, since the new code must somehow come to blockchain. So even if we are only upgrading a facet, a deployment must take place at some point. And I think it sensible to do it here and not have a separate procedure if it's only upgraded.